### PR TITLE
Support soft target in softmax_cross_entropy

### PR DESCRIPTION
**Pull request by anaruse**
This PR aims to support "soft target" in softmax_cross_entropy.

Current softmax_cross_entropy implementation support "hard target" but does not support "soft target" that is becoming popular as a method to mitigate over-fitting. This PR allows users to use "soft target" in softmax_cross_entropy as follows.

    soft_target_loss = F.softmax_cross_entropy(x, t, soft_target=soft_target)

The soft target loss is KL divergence.

### CODE REVIEW

**Comment by anaruse** [comment]
Or, would it be better to implement this as different function such as softmax_kl_divergence?

**Comment by anaruse** [comment]
I've fixed the PR, so that it uses argument `t` for both hard and soft target. Whether it is hard or soft target is determined by ndim and shape of `x` and `t`.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 1]
```diff
@@ -572,4 +572,107 @@ def test_consistency_gpu_never(self):
...
+    'reduce': ['mean', 'no'],
+})))
+class TestSoftTargetCompareToHard(BaseSoftTarget, unittest.TestCase):
+
+    def setUp(self):
+        BaseSoftTarget.setUp(self)
+        t_hard_shape = (self.nb,) + self.shape[1:]
+        self.t_hard = numpy.random.randint(
+            0, self.shape[0], t_hard_shape).astype(numpy.int32)
+        t = numpy.zeros(numpy.prod(self.x.shape)).astype(self.dtype)
```
Comment (beam2d) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]:
```suggestion
        t = numpy.zeros(self.x.size).astype(self.dtype)
```

**Comment by anaruse** [comment]
Thanks for your comments. I've just fixed the branch based on your feedback. Please review it again.

**Comment by beam2d** [comment]
Thank you for the updates. Looks good to me. Could you add a description of the soft target support to the docstring?

**Comment by toslunar** [comment]
Could you resolve merge conflicts?

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 2]
```diff
@@ -72,14 +74,22 @@ def check_type_forward(self, in_types):
...
+                t_type.ndim == x_type.ndim - 1,
+                x_type.shape[0] == t_type.shape[0],
+                x_type.shape[2:] == t_type.shape[1:],
+            )
+        else:
+            # assume t is soft_target
+            type_check.expect(
+                x_type.dtype.kind == 'f',
+                t_type.dtype.kind == 'f',
+                t_type.ndim == x_type.ndim,
```
Comment (toslunar) [inline chainer/functions/loss/softmax_cross_entropy.py]: The following line covers this check.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 3]
```diff
@@ -681,4 +681,107 @@ def test_consistency_gpu_never(self):
...
+        self.check_backward(numpy)
+
+    @attr.gpu
+    def test_backward_gpu(self):
+        self.check_backward(cuda.cupy)
+
+
+@testing.parameterize(*(testing.product({
+    'nb': [1, 2, 4],
+    'shape': [(3,), (3, 2), (3, 2, 2)],
```
Comment (toslunar) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Could you add `()` to the shapes?

Comment (anaruse) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Well, is there a use case when shape is () ?

Comment (anaruse) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Actually, this test code does not work well with shape = ().

**Comment by beam2d** [comment]
Bump. Sorry for the late response. Could you resolve the conflict again?

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py (hunk 3 above)
Comment (beam2d) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: I thought there should be at least one axis in this shape following the documentation of the original hard-target softmax cross entropy.

**Comment by anaruse** [comment]
I've just resolved the conflicts with the master branch. Could you check it?

**Comment by beam2d** [comment]
Sorry for the late reply. The code looks good to me, but I have a bit concern on the naming now. The function named "cross entropy" that computes KL divergence is confusing and it will quite likely be misused. You asked above if it is better to name it `softmax_kl_divergence`, and that sounds better to me, but I also think it's overkill to add a new function, based on the current implementation. How about adding an option to turn on/off the negative entropy term (i.e., switching between cross-entropy mode and KL divergence mode)? The option only affects the soft target case because the cross-entropy and KL divergence are the same for the hard target case.

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 4]
```diff
@@ -223,21 +244,32 @@ def forward_gpu(self, inputs):
             ret = ret.reshape(t.shape)
         return ret,
 
+    def _soft_target_loss(self, xp, x, t, log_y):
+        kl_d = xp.sum(t * (xp.log(t + self.eps) - log_y), axis=1)
```
Comment (toslunar) [inline chainer/functions/loss/softmax_cross_entropy.py]: To compute the cross entropy, 
```suggestion
        ____ = -xp.sum(t * log_y, axis=1)
```

**Comment by toslunar** [comment]
BTW, `chainer.distributions.Categorical` is available.

**Comment by anaruse** [comment]
All right, I added an option 'soft_target_loss' so that you can opt which loss calculation method to use for soft target loss: 'cross-entropy' or 'kl-divergence'.  What would you think about this option?

**Comment by beam2d** [comment]
Thanks for the fix!
CI, test this please.

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 5]
```diff
@@ -223,21 +246,35 @@ def forward_gpu(self, inputs):
             ret = ret.reshape(t.shape)
         return ret,
 
+    def _soft_target_loss(self, xp, x, t, log_y):
+        if self.soft_target_loss == 'kl-divergence':
+            ret = xp.sum(t * (xp.log(t + self.eps) - log_y), axis=1)
+        else:
+            ret = -xp.sum(t * log_y), axis=1)
```
Comment (beam2d) [inline chainer/functions/loss/softmax_cross_entropy.py]: It looks there is a syntax error here.

Comment (anaruse) [reply chainer/functions/loss/softmax_cross_entropy.py]: Ah, sorry, I was careless..  will fix it soon.

File: chainer/functions/loss/softmax_cross_entropy.py (hunk 4 above)
Comment (anaruse) [reply chainer/functions/loss/softmax_cross_entropy.py]: Now, I'm wondering if this is correct, since it fails at the test below. 
https://github.com/chainer/chainer/blob/1ab26322767bec5fe9b0c6b5da702b95cb1ef3bb/tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py#L618
This test uses output of softmax as soft target label, so output of softmax_cross_entropy becomes zero or almost zero. But it becomes non-zero when 'cross-entropy' is used as soft target loss calculation.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by beam2d** [comment]
It looks the test is still failing. Could you check it?

**Comment by anaruse** [comment]
The CI test fails at `TestSoftTargetExpectNearZero` (https://github.com/chainer/chainer/blob/f84721460f96e4074135ae28970f8ad5e4ceb734/tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py#L621) when `'cross-entropy'` is used to compute soft target loss.

The test above expects a loss value to be zero, but it becomes non-zero when `'cross-entropy'` is used. I think the cause is that this test is not appropriate for `'cross-entropy'` or computation method  of `'cross-entropy'` is not correct.

What would you think on this?

**Comment by anaruse** [comment]
Sorry for being very late.
I fixed an issue of unit test `TestSoftTargetExpectNearZero` when `cross-entropy` is selected as soft target loss calculation by dividing the test to two tests, one for `kl-divergence` and another for `cross-entropy`. Perhaps there is no problem remaining. Could you check this again?

**Comment by beam2d** [comment]
Jenkins and flexCI, test this please.

File: chainer/functions/loss/softmax_cross_entropy.py (hunk 4 above)
Comment (toslunar) [reply chainer/functions/loss/softmax_cross_entropy.py]: Done in 1ab26322767bec5fe9b0c6b5da702b95cb1ef3bb.

**Comment by beam2d** [comment]
Thank you!!!

**Comment by anaruse** [comment]
Thank you for merging the PR !
